### PR TITLE
Add unnecessary_escapes_in_string

### PR DESCRIPTION
**Pull request by a14n**
**BAD:**
```
'this string contains 2 \"double quotes\" ';
"this string contains 2 \'single quotes\' ";
```
**GOOD:**
```
'this string contains 2 "double quotes" ';
"this string contains 2 'single quotes' ";
```

### CODE REVIEW

File: lib/src/rules/unnecessary_escapes_in_string.dart [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:meta/meta.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Remove unnecessary backslash in strings.';
```
Comment (bwilkerson) [inline lib/src/rules/unnecessary_escapes_in_string.dart]: nit: "backslash" --> "backslashes"

**Comment by a14n** [comment]
Interestingly this lint catched some issue in regexs on Flutter codebase.

File: test/rules/unnecessary_escapes_in_string.dart [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `pub run test -N unnecessary_escapes_in_string`
+
+f(o){
```
Comment (bwilkerson) [inline test/rules/unnecessary_escapes_in_string.dart]: It would be good to add at least one example of a raw string and at least one example using a multi-line string.

Comment (a14n) [reply test/rules/unnecessary_escapes_in_string.dart]: I don't know how to mark a `// LINT` with multiline string :-(

Comment (a14n) [reply test/rules/unnecessary_escapes_in_string.dart]: done for raw string

File: lib/src/rules/unnecessary_escapes_in_string.dart (hunk 1 above)
Comment (a14n) [reply lib/src/rules/unnecessary_escapes_in_string.dart]: Done

File: test/rules/unnecessary_escapes_in_string.dart (hunk 2 above)
Comment (bwilkerson) [reply test/rules/unnecessary_escapes_in_string.dart]: Multiline strings aren't required to contain end of line markers, so you could do something like
```
f('''_\'_'''); // LINT
f('''_\''''); // OK
f('''\'_'''); // LINT
f('''_\"_'''); // LINT

f("""_\"_"""); // LINT
f("""_\""""); // OK
f("""\"_"""); // LINT
f("""_\'_"""); // LINT
```

Comment (a14n) [reply test/rules/unnecessary_escapes_in_string.dart]: Sure my bad :)

Comment (a14n) [reply test/rules/unnecessary_escapes_in_string.dart]: Actually multiline strings are a little harder to handle.

```dart
f('''_\'_'''); // LINT
f('''_'\''_'''); // OK or it cause error
```

Comment (a14n) [reply test/rules/unnecessary_escapes_in_string.dart]: For now `'''_\'_'''` don't trigger a lint. Only `'''_\"_'''` triggers a lint.

Comment (bwilkerson) [reply test/rules/unnecessary_escapes_in_string.dart]:
```
f('''_'\''_'''); // OK or it cause error
```
That would need to be OK because removing the backslash would result in parse errors.

> For now `'''_\'_'''` don't trigger a lint.

I think that's reasonable, given the complexity of multi-line strings.

Comment (bwilkerson) [reply test/rules/unnecessary_escapes_in_string.dart]: I think it would still be good to add the test cases with comments explaining the desired outcome in cases where the rule doesn't currently produce that outcome. If nothing else we should add the cases that are handled correctly.

Comment (a14n) [reply test/rules/unnecessary_escapes_in_string.dart]: Finally I made the changes for multiline strings.

File: lib/src/rules/unnecessary_escapes_in_string.dart [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  @override
+  void visitStringInterpolation(StringInterpolation node) {
+    for (var element in node.elements.whereType<InterpolationString>()) {
+      visitLexeme(
+        element.contents,
+        isSingleQuoted: node.isSingleQuoted,
+        isMultiline: node.isMultiline,
+        // TODO(a14n): should be the following line but the values look buggy
+        // contentsOffset: element.contentsOffset,
+        // contentsEnd: element.contentsEnd,
```
Comment (a14n) [inline lib/src/rules/unnecessary_escapes_in_string.dart]: I faced strange behaviour with the above code. It looks like a bug to me in the returned values.

For instance `"\'$f"`:

```
node:"\'$f"
node.offset:286
element:"\'
element.offset:286
element.contentsOffset:287   // OK
element.contentsEnd:288      // KO should be 289
element.end:289              // OK
element:"
element.offset:291           // OK
element.contentsOffset:292   // KO should be 291
element.contentsEnd:292      // KO should be 291
element.end:292              // OK
```

@bwilkerson WDYT?

Comment (bwilkerson) [reply lib/src/rules/unnecessary_escapes_in_string.dart]: I agree, those values look wrong.

File: lib/src/rules/unnecessary_escapes_in_string.dart [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    final lexeme = token.lexeme
+        .substring(contentsOffset - token.offset, contentsEnd - token.offset);
+    for (var i = 0; i < lexeme.length; i++) {
+      var current = lexeme[i];
+      var escaped = false;
+      if (current == r'\') {
+        escaped = true;
+        i += 1;
+        current = lexeme[i];
```
Comment (bwilkerson) [inline lib/src/rules/unnecessary_escapes_in_string.dart]: `i` could be past the end of the lexeme at this point.

File: test/rules/unnecessary_escapes_in_string.dart (hunk 2 above)
Comment (bwilkerson) [reply test/rules/unnecessary_escapes_in_string.dart]: Thanks! I really was only concerned with not forgetting the corner cases, but it's great to have them handled.

File: lib/src/rules/unnecessary_escapes_in_string.dart (hunk 4 above)
Comment (a14n) [reply lib/src/rules/unnecessary_escapes_in_string.dart]: For non raw string I don't see how this could happen.

Comment (bwilkerson) [reply lib/src/rules/unnecessary_escapes_in_string.dart]: Oh, right. I missed the fact that the last character of the content can't be a backslash because that would cause the terminating quote to be part of the string. Sorry about that.

File: lib/src/rules/unnecessary_escapes_in_string.dart (hunk 3 above)
Comment (pq) [reply lib/src/rules/unnecessary_escapes_in_string.dart]: @a14n: maybe open a bug on analyzer to track?  (fyi @stereotype441)

**Comment by a14n** [comment]
> Is this good to land?

Yes on my side.

**Comment by pq** [comment]
Thanks!

File: lib/src/rules/unnecessary_escapes_in_string.dart (hunk 3 above)
Comment (a14n) [reply lib/src/rules/unnecessary_escapes_in_string.dart]: Filed dart-lang/sdk#40495
